### ISSUE TITLE
Path query tests fail when SQLITE_LIKE_DOESNT_MATCH_BLOBS

### ISSUE THREAD

**Issue by stefanor** [beetbox/beets#2172]
Related to #1587.

sqlite 3.10 has an option `SQLITE_LIKE_DOESNT_MATCH_BLOBS` [docs](https://sqlite.org/compile.html#like_doesnt_match_blobs) that breaks the path handling logic in beets. It's a performance optimisation, because LIKE doesn't really seem applicable to blobs, and has historically had bugs. Debian (and therefore Ubuntu) are building sqlite with this option as of 3.10.0-1, so you can see this in stretch and xenial.

`test_query` failures resulting from it:

```
======================================================================
FAIL: test_case_sensitivity (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 538, in test_case_sensitivity
    self.assert_items_matched(results, [u'path item', u'caps path'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item', u'caps path']

Second list contains 2 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item', u'caps path']

======================================================================
FAIL: test_escape_backslash (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 524, in test_escape_backslash
    self.assert_items_matched(results, [u'with backslash'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'with backslash']

Second list contains 1 additional elements.
First extra element 0:
u'with backslash'

- []
+ [u'with backslash']

======================================================================
FAIL: test_escape_percent (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 514, in test_escape_percent
    self.assert_items_matched(results, [u'with percent'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'with percent']

Second list contains 1 additional elements.
First extra element 0:
u'with percent'

- []
+ [u'with percent']

======================================================================
FAIL: test_escape_underscore (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 504, in test_escape_underscore
    self.assert_items_matched(results, [u'with underscore'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'with underscore']

Second list contains 1 additional elements.
First extra element 0:
u'with underscore'

- []
+ [u'with underscore']

======================================================================
FAIL: test_nonnorm_path (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 457, in test_nonnorm_path
    self.assert_items_matched(results, [u'path item'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item']

Second list contains 1 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item']

======================================================================
FAIL: test_parent_directory_no_slash (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 425, in test_parent_directory_no_slash
    self.assert_items_matched(results, [u'path item'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item']

Second list contains 1 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item']

======================================================================
FAIL: test_parent_directory_with_slash (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 433, in test_parent_directory_with_slash
    self.assert_items_matched(results, [u'path item'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item']

Second list contains 1 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item']

======================================================================
FAIL: test_path_exact_match (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 417, in test_path_exact_match
    self.assert_items_matched(results, [u'path item'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item']

Second list contains 1 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item']

======================================================================
FAIL: test_slashed_query_matches_path (test.test_query.PathQueryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 465, in test_slashed_query_matches_path
    self.assert_items_matched(results, [u'path item'])
  File "/build/beets-TC7EW2/beets-1.3.19/test/test_query.py", line 81, in assert_items_matched
    self.assertEqual([i.title for i in results], titles)
AssertionError: Lists differ: [] != [u'path item']

Second list contains 1 additional elements.
First extra element 0:
u'path item'

- []
+ [u'path item']

----------------------------------------------------------------------
Ran 114 tests in 3.037s

FAILED (SKIP=1, failures=9)
```

**Comment by sampsyo** [comment]
Thanks for the thorough documentation. It looks like we need to avoid using LIKE for this purpose.

**Comment by jrobeson** [comment]
@sampsyo: so what's the alternative to using `LIKE`?

**Comment by stefanor** [comment]
`left(...) == ...` ?

**Comment by jrobeson** [comment]
haha! yeah sqlite does have functions doesn't it :) 

I'm going to nominate this as a blocker bug for the next release.

**Comment by jrobeson** [comment]
@stefanor : sqlite doesn't have `left()`!.  I've submitted a PR that passes the tests in #2174. Wait for @sampsyo's ok before deciding to add this patch to your package.

**Comment by jrobeson** [comment]
@stefanor : can you verify that https://github.com/beetbox/beets/pull/2176 solves this for you?

**Comment by jrobeson** [comment]
a fix was merged in #2176

**Comment by stefanor** [comment]
Sorry I didn't get around to that in time.
